### PR TITLE
Replace `String` with `SmolStr` for nodes `Identifier` and `ExprName`

### PR DESCRIPTION
**Pull request by LaBatata101**
## Summary

Changes the type of the `id` field in the `Identifier` and `ExprName` nodes from `String` to `SmolStr`.  This is a required change for the new parser.

This PR is part of #9152 

## Test Plan
`cargo test --lib`

### CODE REVIEW

File: crates/ruff_linter/src/rules/pyflakes/rules/strings.rs [hunk 1]
```diff
@@ -747,7 +747,11 @@ pub(crate) fn string_dot_format_extra_named_arguments(
     let missing: Vec<(usize, &str)> = keywords
         .enumerate()
         .filter_map(|(index, keyword)| {
-            if summary.keywords.contains(keyword.as_ref()) {
+            if summary
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: This is unrelated to your change and doesn't need to be fixed as part of this PR: @charliermarsh should we use `HashSet`s instead of `Vec` for these sets (convert from `Vec` to `Set` when converting from `Options` to `Settings`)? Or is the reason we don't do this because we only use `Options` for plugin specific configurations.

File: crates/ruff_linter/src/rules/pylint/rules/no_method_decorator.rs [hunk 2]
```diff
@@ -131,7 +131,7 @@ fn get_undecorated_methods(
 
                         if let Expr::Name(ast::ExprName { id, .. }) = &arguments.args[0] {
                             if target_name == *id {
-                                explicit_decorator_calls.insert(id.clone(), stmt.range());
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/no_method_decorator.rs:134]: GitHub doesn't allow me to make a suggestion on the entire code. We can avoid calling `to_string` here by changing `explicit_decorator_calls` to `HashMap<&str, TextRange>`. 

The `id.to_string` call on linne 128 also seems unnecessary (we can use the `SmolStr` directly or use `as_str`)

File: crates/ruff_python_ast/src/nodes.rs [hunk 3]
```diff
@@ -1739,7 +1740,7 @@ impl From<ExprStarred> for Expr {
 #[derive(Clone, Debug, PartialEq)]
 pub struct ExprName {
     pub range: TextRange,
-    pub id: String,
+    pub id: SmolStr,
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/nodes.rs:1743]: How did you decide where to use `SmolStr`? Should `ExprIpyEscapeCommand`, `StringLiteral`, and `FStringElement` use `SmolStr` too or is it intentional that they do not because they're less likely to be short?

**Review by MichaReiser — COMMENTED** [review]
Thank you. 

The code changes look good to me. The two main questions that I have are:

* How did you decide on `smol_str` over any of the other small string crates? For example, [`compact_str`](https://crates.io/crates/compact_str) supports up to 24 bytes and 12 bytes on 32-bit architectures. But I'm unfamiliar with all the crates, so there might be other considerations. The intention isn't that you change the small string crate, but that we document how we made the decision in case we want to re-visit it in the future.
* How did you decide on where to use `SmolStr`? E.g. you use it for Names and Identifiers but not for String literals.

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_linter/src/rules/pyflakes/rules/strings.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: @MichaReiser -- We could change to `Set`, I don't think there's a specific reason for it. But one problem is that if we _do_ use`HashSet<String>`, then we would need to allocate here to convert from `SmolStr` to `String` for the `contains` check, right?

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: Hmm yeah that might be, because `contains` takes `&T` and not `Deref<T>`. Except if we use `SmolStr` everywhere...

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: That is true... I can try it separately.

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: I should not have brought this up. It's probably not worth it, considering that these vectors only contain very few items and hashing might be more expensive than just comparing the length of every string first.

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pyflakes/rules/strings.rs:750]: I think that's why I left them as vectors, because the size of these is typically very small.

**Comment by LaBatata101** [comment]
> How did you decide on `smol_str` over any of the other small string crates? For example, [`compact_str`](https://crates.io/crates/compact_str) supports up to 24 bytes and 12 bytes on 32-bit architectures. But I'm unfamiliar with all the crates, so there might be other considerations. The intention isn't that you change the small string crate, but that we document how we made the decision in case we want to re-visit it in the future.

I haven't investigated much about the different string crates, I found out about `smol_str` when l was looking into rust-analyzer parser source code. I tested in the new parser to see if it had any performance improvement, and it gave a slight performance increase IIRC, that was the reason I chose it 🤣. But it's probably a good idea to do a proper comparison between different string crates in the future.

> How did you decide on where to use `SmolStr`? E.g. you use it for Names and Identifiers but not for String literals.

Since `SmolStr` allocates strings up to 23 bytes long on the stack, and strings in Python are also used for documentation, they can get quite large, so I didn't see much benefit from using `SmolStr` for string literals. However, for Names and Identifiers, which are usually small, and are a common token in the code, using `SmolStr` for them will, probably, save quite a lot of allocations.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Thank for separating this out!

File: crates/ruff_python_ast/src/nodes.rs (hunk 3 above)
Comment (LaBatata101) [reply crates/ruff_python_ast/src/nodes.rs:1743]: I already answered most of this question in another comment, but I didn't mention `ExprIpyEscapeCommand`. I haven't seen many examples of real world use of `ExprIpyEscapeCommand` so I don't know if they are usually short, like `Names` and `Identifiers`. Can't really tell if they are a suitable candidate to use `SmolStr`.

**Comment by charliermarsh** [comment]
Do you mind either merging `main` (there's a type error against new changes), or giving me edit privileges on this branch to clean up prior to merging?

**Comment by LaBatata101** [comment]
> Do you mind either merging `main` (there's a type error against new changes), or giving me edit privileges on this branch to clean up prior to merging?

I think you already have edit privileges on this branch

**Comment by charliermarsh** [comment]
Ah right, thank you, it was an HTTPS vs. SSH thing.

**Comment by BurntSushi** [comment]
It looks like there is almost [no difference](https://codspeed.io/astral-sh/ruff/branches/LaBatata101:smolstr-id) in the micro-benchmarks? Is this change still worth doing?

File: crates/ruff_python_parser/src/soft_keywords.rs [hunk 4]
```diff
@@ -202,9 +202,7 @@ fn soft_to_name(tok: &Tok) -> Tok {
         Tok::Type => "type",
         _ => unreachable!("other tokens never reach here"),
     };
-    Tok::Name {
-        name: name.to_owned(),
-    }
+    Tok::Name { name: name.into() }
```
Comment (MichaReiser) [inline crates/ruff_python_parser/src/soft_keywords.rs:205]:
```suggestion
    Tok::Name { name: ComactString::new_inline(name) }
```

**Comment by MichaReiser** [comment]
> It looks like there is almost [no difference](https://codspeed.io/astral-sh/ruff/branches/LaBatata101:smolstr-id) in the micro-benchmarks? Is this change still worth doing?

That's the observation I made as well when I did this change a while back https://github.com/astral-sh/ruff/pull/6174

I wouldn't expect any changes for all benchmarks other than the lexer because our (old) parser allocates way too much, so any other improvements are lost in the noise. 

I'm surprised that the lexer benchmarks are regressing. But the microbenchmarks might just be too short and exercise the exact same memory pattern for each run, allowing the allocator to re-use the allocations cheaply. 

Or the change indeed is not as impactful as we thought.

**Comment by MichaReiser** [comment]
See https://github.com/astral-sh/ruff/pull/9229 for a comparison (benchmark) to `CompactStr`

**Comment by MichaReiser** [comment]
Copying this here from the Discord conversation with @LaBatata101 

> My current thinking on small-string crates:
> 
> * SmolStr shows too little benefits at the moment to justify adding a new dependency. That's why I wouldn't merge the PR just yet. We can revisit after switching the parser to see if it brings greater benefits (I'm doubtful but performance is hard to predict)
> CompactStr: It improves the lexer performance by about 5%, but the speedups don't manifest in parser or linter improvements. I suspect that it is because our current parser is slow, and that it reduces the number of allocations for an Identifier from 2 to one, which is significant but maybe not significant enough. The API is a bit cumbersome when it comes to comparing with  str, which raises the entry barrier for new contributors (applies in general, yet another string type in Rust for them to understand)
> * Hipstr: As @Labatata101 pointed out, it requires adding lifetimes to Token and all Nodes which is too big a refactor for now.
> 
> I'm leaning towards deferring the refactor until we've switched to the new parser to re-evaluate the benefits. 
It's also worth considering alternatives or combining it with other improvements reducing the number of allocations. For example, we could explore using an Arena for the AST. This greatly reduces the number of allocations (and removes the Drop time), with the result that allocating for Names proportionally becomes more significant (and removing it yields greater performance improvements). 
>
> What I haven't analyzed yet is whether using a small string crate improves Ruff's memory consumption. A reduced memory consumption on its own would be justification enough for me to switch to a small string crate.

@LaBatata101 do you want to analyze the memory consumption or should we move forward without the small string optimisation for now (and close our both PRs?)

**Comment by LaBatata101** [comment]
> do you want to analyze the memory consumption or should we move forward without the small string optimisation for now (and close our both PRs?)

I think it's better to close both of our PRs for now, and do an in-depth analysis after the new parser is merged.
